### ISSUE TITLE
Fix performance issues

### ISSUE THREAD

**Issue by javigomez** [joomla-projects/com_localise#44]
Currently the first load of com_localise starts processes all the .ini causing very poor performance. In some servers you get timeouts.

We need to study the current component and find a solution that makes it more performant. Maybe the solution could be an AJAX view that processes step by step all the things that needs to be done.

**Comment by javigomez** [comment]
After installing 3 languages and trying to load the component I got this:
![screen shot 2014-05-30 at 14 02 18](https://cloud.githubusercontent.com/assets/1375475/3130134/4984e4b0-e7f2-11e3-9654-69582480c0f4.png)

**Comment by javigomez** [comment]
If you reload the page about 3 times you will get the screen again :P

**Comment by mbabker** [comment]
I'm thinking of the same thing with my overhauling of com_patchtester.  I was going to take the indexer from com_finder and use that as a model to create a multi-step processor.

**Comment by javigomez** [comment]
Thanks @mbabker. Not sure how this indexer works. I'm really interested if you could detail the idea.

Now we have not anymore performance issues at the view thanks to @Skullbock and @fastslack fix at #76. This fix removed this calculation:
![screen shot 2014-06-03 at 13 46 01](https://cloud.githubusercontent.com/assets/1375475/3159679/86adb522-eb14-11e3-9218-0d86013bc959.png)
This calculation I think it was not that much important (do you agree @infograf768?) in exchange we solve the performance issues at administrator/index.php?option=com_localise&view=languages.

But the performance issues are still at /administrator/index.php?option=com_localise&view=translations view due to this calculations: 
![screen shot 2014-06-03 at 13 42 30](https://cloud.githubusercontent.com/assets/1375475/3159633/092f52ae-eb14-11e3-8dd0-b03e90b4c142.png)

@ianmacl mentioning you as we also spoke about the issue in case you have advise for us.

Several ideas were shared during the JandBeyond:
- the indexer of michael
- move all file calculations and strings to the database (witch I think is a lot of work and doesn't necessarily looks like the best option if we can find a way to do it with the .ini files)
- using AJAX calls after rendering the view
- @Skullbock suggested that we could try also to use pagination at /administrator/index.php?option=com_localise&view=translations to calculate only the data only for the listed items, for example for 20 items like in the example:
  ![screen shot 2014-06-03 at 13 50 52](https://cloud.githubusercontent.com/assets/1375475/3159726/32ec7f30-eb15-11e3-8e5c-bc0030b72d43.png)

Any ideas are welcomed ^_^

**Comment by mbabker** [comment]
If on the first load, everything is getting processed and causing memory limit or timeout issues, I'd suggest making it so that you have a toolbar button you have to click to process that data.  The way the indexer works, once you click that button, it uses several AJAX requests to break up the index operation into chunks.  It uses at a minimum 3 requests; the first one just gets all the state data together, the next one starts the full operation and the operation step is repeated based on the batch size (default of 50 items, so if you have 240 items then the operation step is repeated 5 times), then a final cleanup step.  State data is stored in the session between requests to not repeat a step in the operation or reprocess data.

**Comment by ianmacl** [comment]
Understanding the issue more, I would definitely use an asynchronous
request to populate that data.  Load up the page, and then create a request
either for each file or chunk of files.  I would probably keep the results
in local storage with an expiry date.  That way you can retrieve it later
without having to make a request unless it is stale.

If you know which files are changing you can also  invalidate it and reload
it to make sure it stays accurate.

Let me know if that makes sense or not.

Ian

**Comment by fastslack** [comment]
The way Michael poses seems a good way, in fact it is what I use in my extensions to handle processing of large amounts of data. I use a table as a "state" reading the contents of the data (in this case would be to read the files) and record this information on the table. As the different steps are happening it call to update the various flags and index ids.

I use only one controller called "ajax" to do this, and it call for different models like "cleanup" or "migrate". I dont know if this is bad but just is fine for make the code simpler. The URL for the step are very simplified,too for example:

```
index.php?option=com_example&format=raw&task=ajax.cleanup
index.php?option=com_example&format=raw&task=ajax.migrate
```

The JS ajax query must to be chained so the next step just start when the current is finished. Mootools and jQuery has parameters to this. Other thing that helps to me to run JS smootly is to add some small delay to each step so give some time to JS to syncronise.

The main problem here is to separate correctly the logic to the different models.

I wish to use "real" asynchronous communication as WebSocket but this requires also create a server in Joomla! which is currently not possible.

**Comment by infograf768** [comment]
Whatever the solution chosen, we absolutely need the Translated column as it tells to the user immediately which files need attention.

**Comment by javigomez** [comment]
sounds good. @fastslack do you have time to do a pull with a proof of concept? if afterwards we want to rename the method or whatever we can do it later. But let's have something that works.

Once we have the AJAX working we can try to store the results as @ianmacl  said. But let's work step by step.

^_^

**Comment by fastslack** [comment]
I ll try to get the JS stuff working with the steps table for next week. I ll need some help to move the models with the load processes because im not acquainted with the code.

**Comment by javigomez** [comment]
Hi @fastslack, did you had the chance to work on some proof of concept? thanks.

**Comment by fastslack** [comment]
@javigomez Yes I'm working on it. Sorry for the delay but there were many things to attend last week. Surely in the next few days I'll have something to show.

On the other hand, I found an error regarding this topic. Please check the PR b52b44d that solve a bug with # __assets table that whenever we do a purge of the database, does not delete the rows related into this table. This makes that this table increase and affect the performance of this component and also the entire Joomla! installation.
